examples/rl/rllib/rllib.py

```diff
@@ -126,69 +196,95 @@ def main(
-    analysis = tune.run(
-        "PG",
-        name=experiment_name,
-        stop={"time_total_s": time_total_s},
-        checkpoint_freq=1,
-        checkpoint_at_end=True,
-        local_dir=str(result_dir),
-        resume=resume_training,
-        restore=checkpoint,
-        max_failures=3,
-        num_samples=num_samples,
-        export_formats=["model", "checkpoint"],
-        config=tune_config,
-        scheduler=pbt,
-    )
+    ## Approach 2
+    from ray.rllib.algorithms.algorithm import Algorithm
+
+    algo = algo_config.build()
+    if checkpoint is not None:
+        Algorithm.load_checkpoint(algo, checkpoint=checkpoint)
+    result = {}
+    current_iteration = 0
+    checkpoint_iteration = checkpoint_num or 0
+
+    try:
+        while result.get("time_total_s", 0) < time_total_s:
+            result = algo.train()
+            print(f"======== Iteration {result['training_iteration']} ========")
+            print(result, depth=1)
+
+            if current_iteration % checkpoint_freq == 0:
+                checkpoint_dir = get_checkpoint_dir(checkpoint_iteration)
+                print(f"======= Saving checkpoint {checkpoint_iteration} =======")
+                algo.save_checkpoint(checkpoint_dir)
+                checkpoint_iteration += 1
+            current_iteration += 1
+        algo.save_checkpoint(get_checkpoint_dir(checkpoint_iteration))
+    finally:
+        algo.save(get_checkpoint_dir("latest"))
+
+    algo.stop()
 
-    print(analysis.dataframe().head())
+    ## Approach 3
+    # from ray import air
+    # run_config = air.RunConfig(
+    #     name=experiment_name,
+    #     stop={"time_total_s": time_total_s},
+    #     callbacks=[Callbacks],
+    #     storage_path=result_dir,
+    #     checkpoint_config=air.CheckpointConfig(
+    #         num_to_keep=3,
+    #         checkpoint_frequency=checkpoint_freq,
+    #         checkpoint_at_end=True,
+    #     ),
+    #     failure_config=air.FailureConfig(
+    #         max_failures=3,
+    #         fail_fast=False,
+    #     ),
+    #     local_dir=str(result_dir),
+    # )
+    # tune_config = tune.TuneConfig(
+    #     metric="episode_reward_mean",
+    #     mode="max",
+    #     num_samples=num_samples,
+    #     scheduler=pbt,
+    # )
+    # tuner = tune.Tuner(
+    #     "PPO",
+    #     param_space=algo_config,
+    #     tune_config=tune_config,
+    #     run_config=run_config,
+    # )
 
-    best_logdir = Path(analysis.get_best_logdir("episode_reward_max", mode="max"))
-    model_path = best_logdir / "model"
+    # results = tuner.fit()
+    # # Get the best result based on a particular metric.
+    # best_result = results.get_best_result(metric="episode_reward_mean", mode="max")
 
-    copy_tree(str(model_path), save_model_path, overwrite=True)
-    print(f"Wrote model to: {save_model_path}")
+    # # Get the best checkpoint corresponding to the best result.
+    # best_checkpoint = best_result.checkpoint
```

Comment:
I have not finished testing this approach but it appears to be the modern way of working with `Tune` and more flexible than the previous two. It uses `Tune` through the `Tuner` interface. I had a bit of difficulty figuring out where the algorithm config goes.

The examples appear to show it goes in `param_space` but that is an unclear argument.